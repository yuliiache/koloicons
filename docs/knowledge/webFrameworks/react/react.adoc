== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







