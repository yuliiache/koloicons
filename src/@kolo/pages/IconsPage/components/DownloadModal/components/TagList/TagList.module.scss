@import "@kolo/design";

.iconTags {
  display: flex;
  flex-wrap: wrap;
  font-size: 14px;
  line-height: 1.43;
  color: $GreyColor;
  text-transform: capitalize;
  margin: 12px 0 24px 0;
  cursor: pointer;
  outline: none;
  width: 100%;
  gap: 3px;
  padding: 0;
}

.tagItem {
  display: flex;

  &:hover {
    color: $BlueColor;
  }
}
