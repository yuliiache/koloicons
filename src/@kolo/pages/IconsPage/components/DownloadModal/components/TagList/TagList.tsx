import {SEARCH_QUERY} from '@kolo/constants/constants';
import {addQueryParams} from '@kolo/services/helpers/urls';
import {TooltipPlacement} from '@kolo/uiKit/Tooltip/constants';
import Tooltip from '@kolo/uiKit/Tooltip/Tooltip';
import {ApiEndpoint} from 'constants/ApiEndpoint';
import {useLocation, useNavigate} from 'react-router-dom';

import styles from './TagList.module.scss';

export const TagList = ({tags}: {tags?: string[]}) => {
  const navigate = useNavigate();
  const {search} = useLocation();
  const lastTag = tags && tags.length - 1;

  const tagQueryHandle = (tagName: string) => {
    navigate({
      pathname: ApiEndpoint.ICONS,
      search: addQueryParams(search, {
        [SEARCH_QUERY]: tagName,
      }),
    });
  };

  return (
    <ul className={styles.iconTags}>
      {tags &&
        tags.map((tag, index) => (
          <Tooltip
            label="Search by tag"
            key={tag}
            placement={TooltipPlacement.BOTTOM}
          >
            <li
              className={styles.tagItem}
              onClick={() => tagQueryHandle(tag)}
              role="link"
            >{`#${tag}${index !== lastTag ? ',' : ''}`}</li>
          </Tooltip>
        ))}
    </ul>
  );
};
export default TagList;
