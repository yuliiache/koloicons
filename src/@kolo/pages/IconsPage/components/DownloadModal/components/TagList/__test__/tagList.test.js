import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import {render} from '@testing-library/react';

import TagList from '../TagList';

describe('TagList component', () => {
  const tags = ['Accounting', 'Accountant', 'Calculate', 'Calculation'];
  it('renders correctly and matches snapshot', () => {
    const {asFragment} = render(
      <MemoryRouter>
        <TagList tags={tags} />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
