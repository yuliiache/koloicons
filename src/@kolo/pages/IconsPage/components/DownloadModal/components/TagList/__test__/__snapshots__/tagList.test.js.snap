// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagList component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <ul
    class="iconTags"
  >
    <li
      class="tagItem"
      role="link"
    >
      #Accounting,
    </li>
    <li
      class="tagItem"
      role="link"
    >
      #Accountant,
    </li>
    <li
      class="tagItem"
      role="link"
    >
      #Calculate,
    </li>
    <li
      class="tagItem"
      role="link"
    >
      #Calculation
    </li>
  </ul>
</DocumentFragment>
`;
